specs/vep-1493-vault-integration/README.md

```diff
@@ -0,0 +1,223 @@
+Service Data Base. This enhancement will allow users to securely store secrets/passwords/credentials/tokens/data into
+a vault.
+
+
+## Requirements and goals
+
+Goals:
+- provide the ability to optionally configure a Vault instance in the Control Service for storing secrets
+- change the APIs and SDK so that users can set and retrieve secrets and data jobs can use them during runtime
+
+Non-Goals:
+- automatic provisioning of a Vault instance as part of the VDK-CS installation
+- automatic migration of existing secrets to Vault
+
+## High-level design
+
+In order to implement the proposed change, we should make the following changes - highlighted in the diagram below:
+- provide the ability to optionally configure a Vault instance in the Control Service for storing secrets
+- change the CS to store properties marked as secret into the configured Vault
+- enhance vdk-sdk to allow for users to store secrets for their data jobs
+- during runtime, data jobs can retrieve the secrets
+
+![high_level_design.png](diagrams/high_level_design.png)
+
+## API design
+
+Currently, data job properties are stored and read in bulk. We are going to introduce similar APIs for storing and
+reading secrets.
+
+#### Storing/retrieving secrets Control Service API
+
+Introduce new API methods for storing and retrieving secrets under:
+[/data-jobs/for-team/{team_name}/jobs/{job_name}/deployments/{deployment_id}/properties]()
+
+```yaml
+  '/data-jobs/for-team/{team_name}/jobs/{job_name}/deployments/{deployment_id}/secrets':
+    summary: |
+      Data Job Secrets API.
+    get:
+      tags:
+        - Data Jobs Secrets
+      responses:
+        '200':
+          $ref: '#/components/responses/Secrets'
+      operationId: dataJobSecretsRead
+      summary: Get Data Job secrets.
+      description: <i>(Introduced in v1.0)</i>
+    put:
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/DataJobSecrets'
+        required: true
+      tags:
+        - Data Jobs Secrets
+      responses:
+        '201':
+          description: Created successfully.
+        '204':
+          description: Updated successfully.
+      operationId: dataJobSecretsUpdate
+      summary: Update Data Job secrets.
+      description: <i>(Introduced in v1.0)</i>
+    parameters:
+      -  name: team_name
+         description: Team Name
+         schema:
+           type: string
+         in: path
+         required: true
+      - name: job_name
+        description: Data Job Name.
+        schema:
+          type: string
+        in: path
+        required: true
+      - name: deployment_id
+        description: Data job deployment id.
+        schema:
+          $ref: '#/components/schemas/DataJobDeploymentId'
+        in: path
+        required: true
+```
+
+When storing secrets, the request body is expected to provide a JSON object, containing all secrets:
+
+```json
```

Comment:
Have you considered the following format?

```json
{
  "secrets": [
    {"name": "host", "value": "host1.com"},
    {"name": "user", "value": "user1"},
    {"name": "pass", "value": "1234"}
  ]
}
```

Reply:
Yes, however it is different than the current properties format which means that either we'll have two different interfaces for the users, or we'll have to convert them in the code...

Reply:
I agree with Dako. If we want to change the format, we should do it for both.

[Vault supports JSON format](https://developer.hashicorp.com/vault/docs/commands/kv/put). I suggest we use it And store in in the same format as we do in Properties (JSON).  I don't believe we should limit what structure the user choose to store their secrets in (only string). 


This reminds me that we should consider size limits (I read that a common limitation for many vault storage backends is around 512 KB per key-value pair in Vault). That would go in Detail design section. But it came to mind now.

Reply:
As you said the interface will remain the same, we will convert them in the code. I am not saying that this is better or not, but if you think it is better I don't see any problem to proceed with it.

Reply:
In my opinion, we are a bit limited with this approach since we don't have the ability to update only one of the properties/secrets (although it is possible to update a particular field in the JSON column in the database, it is hard to be implemented). In the case of a data job with a lot of properties/secrets that updates only one property/secret, we need to load all properties/secrets and then send them back to the CS, which will be slightly inefficient.

Reply:
I think that we need to support CRUD operations in the API for a single secret.
Why because when updating the secrets in bulk, you need to retrieve the other values and they offer the user in  VDK/UI to edit all values.

This was an issue with the UI implementation when it was started in the summer.

Reply:
I believe this also need to be added for properties, as the UI to be consistent or very close.

Reply:
in the VEP, it is suggested to support a command "vdk secrets --set "my-password" - which I suggest - single secret update, how is this going to be supported if the API supports bulk edit?
{
  "secrets": [
    {"name": "host", "value": "host1.com"},
    {"name": "user", "value": "user1"},
    {"name": "pass", "value": "1234"}
  ]
}

Reply:
> In my opinion, we are a bit limited with this approach since we don't have the ability to update only one of the properties/secrets

I am not sure we are that limited 
By Contract ([API defintion](https://github.com/vmware/versatile-data-kit/blob/main/projects/control-service/projects/model/apidefs/datajob-api/api.yaml#L1222) and [Job Input SDK definition](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/api/job_input.py#L32) ) the properties are dictionary where the key must be string, the value can be primtiive type, list or dict. 


In other words Properties type is Dict[str, PropertyValue]) where PropertyValue = Union[int, float, str, list, dict, None] ([here](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/job_properties/datajobs_service_properties.py#LL70C46-L70C71))
and [here](https://github.com/vmware/versatile-data-kit/blob/main/projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/properties/controller/DataJobsPropertiesController.java#LL36C61-L36C71)


So it will not be a breaking change to introduce:  '/data-jobs/for-team/{team_name}/jobs/{job_name}/deployments/{deployment_id}/properties/top_level_key/ = PropertyValue

should we decide to Impl wise it would still need to fetch all properties and update them. But it's db - to service communication now behind interface. 
As Properties DB model is implemntation detail we can always change it if we are not happy.

Reply:
I've added a couple of points in the Alternatives section for clarity.

Reply:
Although I think it would be even better to improve the API format, it sounds outside the scope of the VEP. Let's be consistent with the Properties API.